pkg/metrics/collector/pod_ip_metrics.go

```diff
@@ -0,0 +1,174 @@
+		"used_ipv6_addr_count",
+		"Indicates how many IPv6 addresses are in use.",
+		nil, nil,
+	)
+	dualStackCountDesc = prometheus.NewDesc(
+		"dual_stack_count",
+		"Indicates how many pods have dual stack IP addresses.",
+		nil, nil,
+	)
+	dualStackErrorCountDesc = prometheus.NewDesc(
+		"dual_stack_error_count",
+		"Indicates how many pods did not get dual stack IPs erroneously.",
+		nil, nil,
+	)
+	duplicateIpCountDesc = prometheus.NewDesc(
+		"duplicate_ip_count",
+		"Indicates how many pods had more than one IP per family.",
+		nil, nil,
+	)
+)
+
+type podIpMetricsCollector struct {
+	usedIpv4AddrCount uint64
+	usedIpv6AddrCount uint64
+	dualStackCount uint64
+	dualStackErrorCount uint64
+	duplicateIpCount uint64
+}
+
+func init() {
+	registerCollector("pod_ip_metrics", NewPodIpMetricsCollector)
+}
+
+// NewPodIpMetricsCollector returns a new Collector exposing pod IP allocation
+// stats.
+func NewPodIpMetricsCollector() (Collector, error) {
+	return &podIpMetricsCollector{}, nil
+}
+
+func readLine(path string) (string, error) {
+	buf, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+
+	defer func() {
+		if err = buf.Close(); err != nil {
+			glog.Errorf("Error closing file %s: %s", path, err)
+		}
+	}()
+
+	s := bufio.NewScanner(buf)
+	s.Scan()
+	return s.Text(), s.Err()
+}
+
+func (c *podIpMetricsCollector) listIpAddresses(dir string) error {
+	f, err := os.Open(dir)
+	if err != nil {
+		glog.Errorf("Error opening directory %s, %v", dir, err)
+		return err
+	}
+	files, err := f.Readdir(0)
+	if err != nil {
+		glog.Errorf("Error while reading files in directory %v", err)
+		return err
+	}
+
+	podMap := make(map[string]ipFamily)
+	var ipv4Count, ipv6Count, dupIpCount, dualCount, dualErrCount uint64
+	for _, v := range files {
+		if ip := net.ParseIP(v.Name()); ip == nil {
+			// This isn't an IP address file
+			continue
+		}
+		var family ipFamily
+		if strings.Contains(v.Name(), ":") {
+			ipv6Count++
+			family = ipv6
+		} else {
+			ipv4Count++
+			family = ipv4
+		}
+		// Update the map and track IP families only for dual stack clusters
+		fileName := fmt.Sprintf("%s/%s", dir, v.Name())
+		podId, err := readLine(fileName)
+		if err != nil {
+			glog.Errorf("Error reading file %s: %v", fileName, err)
+			continue
+		}
+		f, ok := podMap[podId]
+		if !ok {
+			podMap[podId] = family
+		} else if (f == ipv4 && family == ipv6) || (f == ipv6 && family == ipv4) {
+			podMap[podId] = dual
+		} else {
+			dupIpCount++
+		}
+	}
+	if stackType == dualStack {
+		for _, family := range podMap {
+			if family == dual {
+				dualCount++
+			} else {
+				dualErrCount++
+			}
+		}
+	}
+	c.usedIpv4AddrCount = ipv4Count
+	c.usedIpv6AddrCount = ipv6Count
+	c.dualStackCount = dualCount
+	c.dualStackErrorCount = dualErrCount
+	c.duplicateIpCount = dupIpCount
+	return nil
+}
+
+func (c *podIpMetricsCollector) Update(ch chan<- prometheus.Metric) error {
+	if err := c.listIpAddresses(gkePodNetworkDir); err != nil {
+		return nil
```

Comment:
Humm, probably log the error here instead? I know listIpAddresses() already logs that - but at the first glance it seems like omitting the error without looking into that func.

Reply:
Done